### PR TITLE
Allow all URL schemes in Markdown links by default

### LINKED ISSUE #16721: Matrix URIs aren't rendered as links
- Gitea version (or commit ref): 1.15.0-rc3
- Can you reproduce the bug at https://try.gitea.io:
  - [x] Yes: https://try.gitea.io/Mikaela/gist-manual/src/branch/master/matrix/spaces.md

## Description

Linking to a Matrix URI such as: `[Mikaela.info](matrix:roomid/psumPMeAfzgAeQpXMG:feneas.org?action=join&via=feneas.org&via=sorunome.de&via=the-apothecary.club&via=pirateriot.net&via=matrix.org)` renders as text instead of a clickable link.

<s>Matrix URI scheme is proposed in https://github.com/matrix-org/matrix-doc/blob/master/proposals/2312-matrix-uri.md and is being proposed in https://github.com/matrix-org/matrix-doc/pull/3168.</s>

Update: [Matrix URI scheme is specified in Matrix 1.2](https://spec.matrix.org/latest/appendices/#matrix-uri-scheme) and [recognised by IANA](https://www.iana.org/assignments/uri-schemes/uri-schemes.xhtml).

## Screenshots

* https://gitea.blesmrt.net/mikaela/gist/src/branch/master/matrix/spaces.md
  * image
  * image

### LINKED ISSUE #21146: Alllow all schemes in markdown by default
### Description

Using the main intance https://gitea.com I am unable to insert non-http emails. In particular I am interested to add links to my emails (I use the tool only internally) provided by the [https://camiel.bouchier.be/en/cb_thunderlink](https://camiel.bouchier.be/en/cb_thunderlink) extension. Unfortunately links like `cbthunderlink://somebase64string` are not clickable, even if I explicitely use the longer url notation:

```
See you can't click me [cbthunderlink://somebase64string](cbthunderlink://somebase64string)
```

### Gitea Version

1.18.0+dev-333-g9e0c43777

### Can you reproduce the bug on the Gitea demo site?

Yes

### Log Gist

_No response_

### Screenshots

image

image

### Git Version

_No response_

### Operating System

_No response_

### How are you running Gitea?

Using the instance from https://gitea.com

### Database

_No response_

### PR DESCRIPTION
**Pull request by yardenshoham**
- Closes #21146
- Closes #16721

## :warning: BREAKING :warning:
This changes the default behavior to now create links for any URL scheme when the user uses the markdown form for links (`[label](URL)`), this doesn't affect the rendering of inline links. To opt-out set the `markdown.CUSTOM_URL_SCHEMES` setting to a list of allowed schemes, all other schemes (except `http` and `https`) won't be allowed.

# Before
image

# After
image

### CODE REVIEW

File: modules/markup/sanitizer.go [hunk 1]
```diff
@@ -22,7 +22,10 @@ type Sanitizer struct {
 	init             sync.Once
 }
 
-var sanitizer = &Sanitizer{}
+var (
+	sanitizer     = &Sanitizer{}
+	allowAllRegex = regexp.MustCompile(".+")
```
Comment (silverwind) [inline modules/markup/sanitizer.go:27]: Would `".+:.+"` work as well? Just a tiny bit stricter.

Comment (yardenshoham) [reply modules/markup/sanitizer.go:27]: It doesn't work, the scheme itself doesn't contain `:`

Comment (silverwind) [reply modules/markup/sanitizer.go:27]: Ah, so it's splitting at the first `:`, I suppose.

Comment (silverwind) [reply modules/markup/sanitizer.go:27]: Interestingly, this parsing is different than how the WHATWG URL standard does it, where `protocol` does include the `:`, e.g. `https:`.

**Review by silverwind — APPROVED** [review]

**Comment by silverwind** [comment]
Some unit tests may be nice.

**Comment by lunny** [comment]
Should we allow users to indicate the allowed schemes?

**Comment by yardenshoham** [comment]
I think the sanitizer doesn't have a context

**Comment by yardenshoham** [comment]
How could I make the default of `CUSTOM_URL_SCHEMES` a `*`? I see no equivalent for `rootCfg.Section("markdown").Key("ALLOW_ALL_URL_SCHEMES").MustBool(true)` with a string slice

**Comment by wxiaoguang** [comment]
In `CustomLinkURLSchemes`, if any `scheme` is `*`, then `common.LinkRegex = regexp(".+")`

**Review by wxiaoguang — APPROVED** [review]
It looks good to me as long as it's marked as breaking.

I think it's browser's responsibility to protect users from phishing schemes. Rendering any scheme on a page should be safe.

**Review by silverwind — APPROVED** [review]

**Comment by SISheogorath** [comment]
Uhm, any plans to issue a CVE for this one?

**Comment by delvh** [comment]
Yes, we have reserved CVE-2023-3851 for it.

**Pull request merged**